chain/db/pubsub.go

```diff
@@ -2,95 +2,220 @@ package db
+		return err
+	}
 
-	return subscriberKey.Bytes()
+	return nil
 }
 
-func (cs *ChainStore) Subscribe(subscriber []byte, identifier string, topic string, bucket uint32, duration uint32, meta string, height uint32) error {
-	if duration == 0 {
-		return nil
-	}
+func (ps *pubSub) Deserialize(r io.Reader) error {
+	var err error
 
-	subscriberKey := generateSubscriberKey(subscriber, identifier, topic, bucket)
+	ps.subscriber, err = serialization.ReadVarString(r)
+	if err != nil {
+		return err
+	}
 
-	// PUT VALUE
-	err := cs.st.BatchPut(subscriberKey, []byte(meta))
+	ps.meta, err = serialization.ReadVarString(r)
 	if err != nil {
 		return err
 	}
 
-	err = cs.ExpireKeyAtBlock(height+duration, subscriberKey)
+	ps.expiresAt, err = serialization.ReadUint32(r)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (cs *ChainStore) Unsubscribe(subscriber []byte, identifier string, topic string, bucket uint32, duration uint32, height uint32) error {
-	if duration == 0 {
-		return nil
+func (ps *pubSub) Empty() bool {
+	return len(ps.subscriber) == 0 && len(ps.meta) == 0 && ps.expiresAt == 0
+}
+
+func getTopicId(topic string) []byte {
+	topicHash := sha256.Sum256([]byte(strings.ToLower(topic)))
+	return topicHash[:20]
+}
+
+func getTopicBucketId(topic string, bucket uint32) []byte {
+	topicBucketId := bytes.NewBuffer(nil)
+
+	topicId := getTopicId(topic)
+	topicBucketId.Write(topicId)
+
+	serialization.WriteUint32(topicBucketId, bucket)
+
+	return topicBucketId.Bytes()
+}
+
+func getPubSubId(topic string, bucket uint32, subscriber []byte, identifier string) string {
+	pubSubId := bytes.NewBuffer(nil)
+
+	pubSubId.Write(getTopicBucketId(topic, bucket))
+
+	subscriberHash := sha256.Sum256(append(subscriber, identifier...))
+	pubSubId.Write(subscriberHash[:20])
+
+	return string(pubSubId.Bytes())
+}
+
+func getPubSubCleanupId(height uint32) string {
+	buf := new(bytes.Buffer)
+	_ = serialization.WriteUint32(buf, height)
+	return string(buf.Bytes())
+}
+
+func (sdb *StateDB) getPubSub(id string) (*pubSub, error) {
+	var ps *pubSub
+	var ok bool
+	if ps, ok = sdb.pubSub[id]; !ok {
+		enc, err := sdb.trie.TryGet(append(PubSubPrefix, id...))
+		if err != nil {
+			return nil, err
+		}
+
+		ps = &pubSub{}
+
+		if len(enc) > 0 {
+			buff := bytes.NewBuffer(enc)
+			if err := ps.Deserialize(buff); err != nil {
+				return nil, fmt.Errorf("[getPubSub]Failed to decode state object for pub sub: %v", err)
+			}
+		}
+
+		sdb.pubSub[id] = ps
 	}
 
-	subscriberKey := generateSubscriberKey(subscriber, identifier, topic, bucket)
+	return ps, nil
+}
 
-	// DELETE VALUE
-	err := cs.st.BatchDelete(subscriberKey)
+func (sdb *StateDB) getPubSubCleanup(height uint32) (map[string]struct{}, error) {
+	var psc map[string]struct{}
+	var ok bool
+	if psc, ok = sdb.pubSubCleanup[height]; !ok {
+		enc, err := sdb.trie.TryGet(append(PubSubCleanupPrefix, getPubSubCleanupId(height)...))
+		if err != nil || len(enc) == 0 {
```

Comment:
I see that there is a `len(enc) == 0` here but not in `getPubSub`. Is there a reason for the difference?

Reply:
I think it's a bug (same on in nano pay), fixed